Added `DocketChange` Model & Updated Schema

jcq opened the pull request:
Connects #15137

### Description
This adds new `DocketChange` model as well as updating schema to add `docket_changes` table. These support Docket Switch workflows, and are analogous to `PostDecisionMotion` model for MTV.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] `DocketChange` model added
- [ ] Schema updated to add `docket_changes` table

### Database Changes
*Only for Schema Changes*

* [ ] Timestamps (created_at, updated_at) for new tables
* [ ] Column comments updated
* [ ] ~Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`)~
* [ ] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [ ] Query profiling performed (eyeball Rails log, check bullet and fasterer output)
* [ ] Appropriate indexes added (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [ ] DB schema docs updated with `make docs` (after running `make migrate`)
* [ ] #appeals-schema notified with summary and link to this PR

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,51 @@
+    decision_issues_for_switch.map { |di| di.create_contesting_request_issue!(appeal) }
+  end
+
+  def granted_decision_issues
```
One change from MTV is that these appeals are still active, and the issues that would move are the request issues.

I think you actually want to do this with the currently active request issues, so that will simplify some of this logic. These are the issues that would trigger the "this issue is already under review on an Appeal" ineligibility reason if an intake user tried to add the issue onto another intake.

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,51 @@
+    @granted_decision_issues ||= request_issues.map { |ri| ri.decision_issues.first }
+  end
+
+  def create_granted_decision_issues
```
Related to my previous comment, I don't understand yet what this is for.

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,51 @@
+
+  private
+
+  def granted_issues_present_if_partial
```
Since these appeals are still active, the issues on them could change, so it might be a good idea to capture the request issue IDs on full grants as well.

leikkisa replied:
Realizing the flaw in this comment (since if they edit issues later, it'll be directly on the stream, not involving any copying or anything like that), resolving

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  def create_granted_decision_issues
+    request_issues.map(&:create_granted_decision_issue!)
```
For some reason I can't find this method with search.  Is the idea here that the issues that are being moved to a new docket type would get automatically decided as "granted"?

This may depend on business requirements, but since the request issues are moving, you could probably just close them (or actually move them).  If they already have a decision issue on them, that might impact docket change eligibility, but one option could be to also move the decision issue.

leikkisa replied:
If "docket changed" actually reflects an official decision / disposition type (that would eventually go onto a decision document), then I'd need to adjust some of my comments.

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,51 @@
+
+  # Creates request issues on the new docket contesting the decisions to be granted
+  def create_request_issues_for_switch
+    decision_issues_for_switch.map { |di| di.create_contesting_request_issue!(appeal) }
```
Should `appeal` here be a new appeal stream?

leikkisa commented on db/migrate/20200921191403_create_docket_changes.rb:
```diff
@@ -0,0 +1,17 @@
+class CreateDocketChanges < ActiveRecord::Migration[5.2]
+  def change
+    create_table :docket_changes, comment: "Stores the disposition and associated data for Docket Change motions" do |t|
+      t.belongs_to :appeal, null: false, foreign_key: true
```
I think some comments might be useful for appeal (esp if appeal streams are involved), task, and receipt date.

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class DocketChange < CaseflowRecord
+  include HasSimpleAppealUpdatedSince
```
I'm not sure what this is used for, curious though!

jcq replied:
Will remove!

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  def granted_decision_issues
+    @granted_decision_issues ||= request_issues.map { |ri| ri.decision_issues.first }
```
This would be reliable as long as the request issue didn't have any decision issues already on it before.

leikkisa reviewed (commented):
I have some comments about the request/decision issue relationship (which could depend somewhat on how the business wants these handled).  A quick chat might be useful.

jcq commented:
> I have some comments about the request/decision issue relationship (which could depend somewhat on how the business wants these handled). A quick chat might be useful.

@leikkisa  All very good points! I had renamed various methods, but had otherwise left the logic the same as `PostDecisionMotion` until we knew exactly how we were going to handle things. That said, it doesn't make much sense to leave placeholders like that, so I've gone ahead and updated the various methods to indicate that we'll actually simply be moving the specified request issues from one appeal stream to the other. I'd love to hear your thoughts on that behavior vs keeping (but closing out) the issues on the old one (in addition to creating new ones on the new stream).

jcq commented on db/migrate/20200921191403_create_docket_changes.rb:
```diff
@@ -0,0 +1,18 @@
+      t.datetime "receipt_date", null: false
+      t.string "docket", comment: "The new docket"
+      t.string "disposition", comment: "Possible options are granted, partially_granted, and denied"
+      t.integer "granted_decision_issue_ids", comment: "When a docket change is partially granted, this includes an array of the appeal's decision issue IDs that were selected for the new docket. For full grant, this includes all prior decision issue IDs.", array: true
```
Change this to request issues

leikkisa commented on app/models/docket_change.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def move_granted_request_issues
+    request_issues_for_switch.map { |ri| ri.update!(decision_review: new_docket_stream) }
```
I know this was potentially an open question, but looks good for now.

leikkisa commented on db/migrate/20200921191403_create_docket_changes.rb:
```diff
@@ -0,0 +1,18 @@
+      t.datetime "receipt_date", null: false
+      t.string "docket_type", comment: "The new docket"
+      t.string "disposition", comment: "Possible options are granted, partially_granted, and denied"
+      t.integer "granted_request_issue_ids", comment: "When a docket change is partially granted, this includes an array of the appeal's request issue IDs that were selected for the new docket. For full grant, this includes all prior decision issue IDs.", array: true
```
Small nit on this comment, it says "decision issue IDs" at the end instead of "request issue IDs".

leikkisa commented on spec/models/docket_change_spec.rb:
```diff
@@ -0,0 +1,51 @@
+    cotb_team.add_user(cotb_user)
+  end
+
+  context "#move_granted_request_issues" do
```
Nice, still got some extra functionality in there.

leikkisa reviewed (approved):
Hey JC, very minor nit on the DB comment, but looks good to me! Thanks for the changes.

leikkisa commented on spec/factories/docket_change.rb:
```diff
@@ -0,0 +1,12 @@
+FactoryBot.define do
+  factory :docket_change do
+    old_docket_stream { create(:appeal, docket_type: Constants.AMA_DOCKETS.evidence_submission) }
+    new_docket_stream { create(:appeal, stream_type: "original") }
```
Reminds me that non-original AMA appeals may be a case to prevent these on? Not really sure
